### PR TITLE
Monomorphize in check mode to report (almost) all PMEs

### LINKED ISSUE #99682: 'cargo check' passes but 'cargo build' fails
Consider the following code:
```rust
pub struct G;

pub trait IAmArray{
    const SIZE: usize;
}

impl<T, const N: usize> IAmArray for [T; N]{
    const SIZE: usize = N;
}

impl G{
    pub fn i_am_break_on_different_array_sizes<A, B>(_a: A, _b: B)
        where A: IAmArray,
            B: IAmArray,
    {
        trait CompileAssert{
            const TRIGGER: ();
        }
        impl<A, B> CompileAssert for (A, B)
            where A: IAmArray,
                B: IAmArray,
        {
            const TRIGGER: () = if A::SIZE == B::SIZE {} else {panic!("You must provide arrays of same length")};
        }
        
        let _ = <(A, B) as CompileAssert>::TRIGGER;
    }
}

fn main() {
    G::i_am_break_on_different_array_sizes([0u8;5], [0u32;3]);
}
```
When I do a check-build of this (`rustc --emit=metadata`), it works fine. But when I do a full build (`rustc --emit=link`), it fails:
```
error[E0080]: evaluation of `<([u8; 5], [u32; 3]) as G::i_am_break_on_different_array_sizes::CompileAssert>::TRIGGER` failed
  --> test.rs:23:64
   |
23 |             const TRIGGER: () = if A::SIZE == B::SIZE {} else {panic!("You must provide arrays of same length")};
   |                                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'You must provide arrays of same length', test.rs:23:64
   |
   = note: this error originates in the macro `$crate::panic::panic_2015` which comes from the expansion of the macro `panic` (in Nightly builds, run with -Z macro-backtrace for more info)

note: the above error was encountered while instantiating `fn G::i_am_break_on_different_array_sizes::<[u8; 5], [u32; 3]>`
  --> test.rs:31:5
   |
31 |     G::i_am_break_on_different_array_sizes([0u8;5], [0u32;3]);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error
```
This is a post-monomorphization error. It is a long-known problem but I could not find an issue tracking it.

It seems worth tracking as it is rather surprising, so if we can find a way to fix it, that would be great. It also leads to issues with Miri such as https://github.com/rust-lang/miri/issues/2423.

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #99682

see https://github.com/rust-lang/rust/pull/104087#issuecomment-1409934786 for discussion

r? @ghost

### CODE REVIEW

**Comment by RalfJung** [comment]
What's the "almost" about?

**Comment by oli-obk** [comment]
> What's the "almost" about?

codegen backends can still bail out if they don't support something (like some huge array size limits that are very target dependent)

**Comment by RalfJung** [comment]
We do check for "types can't be larger than 2^48" somewhere on the MIR level already -- so this is a separate check?

Anyway, would be good to have a concrete example and an issue tracking that, but doesn't sound high priority to fix.

**Comment by oli-obk** [comment]
We have tests for

* broken inline assembly
* `[u8; usize::MAX]` (seems to still be a build-fail test after this PR, will need to investigate)
* duplicate symbols via `#[export_name = "dup"]` being used twice on different functions or statics
* invalid uses of the unstable platform intrinsics 🤷 
* various invalid command line arguments
    * invalid `-Cpasses`
    * invalid `-Clinker=asdflkjasdf`
    * ...
* mis-use of `#[target_feature]`

**Comment by RalfJung** [comment]
And const-eval errors, I presume?

**Comment by RalfJung** [comment]
> fixes https://github.com/rust-lang/rust/issues/99682

Turns out we have some duplicates / closely related issues: https://github.com/rust-lang/rust/issues/70923 https://github.com/rust-lang/rust/issues/49292. Are they all fixed?

**Comment by oli-obk** [comment]
cc @tmandry

**Comment by tmandry** [comment]
Thanks for putting this up, I'm very curious to see the perf results!

I think the use of PME here is slightly confusing – if I understand correctly we would still have errors that depend on which generic parameters are used, like the example in https://github.com/rust-lang/rust/pull/104087#issuecomment-1375713665.

**Comment by JakobDegen** [comment]
@oli-obk another implementation strategy that we should consider here is adding a new field to `mir::Body` that lists all the constants that were written into the body by the user. Then, when we monomorphize, we'll continue to do the normal thing, but additionally if the collection does not pick up any of these constants then we'll also mono them in a "const error only mode" (sort of like what you have here).

Part of the reason that I bring this up is because I knew the topic of this issue was a possibility, but thought it was a non-issue specifically because I was under the impression that the `required_consts` field on `Body` already did what I'm describing here. Faults on me for not actually checking that theory

**Comment by RalfJung** [comment]
How is required_consts not already doing that?

The trouble is that we don't have required_fns so we might not transitively monomorphize functions even though they were mentioned by the user.

**Comment by JakobDegen** [comment]
> How is required_consts not already doing that?

Well, what I didn't realize was that required_consts only contains consts that are unevaluated by the time we do borrowck. What I thought happened is that all consts are put into there, specifically to ensure that things like this don't happen

**Comment by RalfJung** [comment]
> Edit: Oh, I realize what the confusion might have been. I was implicitly assuming that if we put a FnDef-typed constant into the list of constants, that we'd then go and actually mono that function (at least enough to go set off any PMEs)

Ah -- no, all we do with that list is evaluate constants, we don't monomorphize any further functions.

**Comment by nbdd0121** [comment]
If we want to make `if const { xxx }` not monomorphize it `xxx` evaluates to false (as suggested in #85836), then we actually would want some constants/functions to not show up in `required_consts`.

**Comment by RalfJung** [comment]
> If we want to make `if const { xxx }` not monomorphize it `xxx` evaluates to false (as suggested in #85836), then we actually would want some constants/functions to not show up in `required_consts`.

That's quite a big new feature that requires an RFC IMO. I don't think there is currently a plan to implement this.

**Comment by JakobDegen** [comment]
Yeah, agreed. That kind of feature would also definitely need explicit Mir support

**Comment by tmandry** [comment]
There's a [lang team meeting](https://github.com/rust-lang/lang-team/issues/195) that will touch on this PR next week (Feb 22).

@oli-obk Do you think you could get it fixed up with a perf run by then? 🙏

**Comment by tmiasko** [comment]
The `should_codegen` check from `should_encode_mir` will probably have to go as well now.

**Comment by oli-obk** [comment]
> The `should_codegen` check from `should_encode_mir` will probably have to go as well now.

I think we can avoid it. We only need to care about generic functions, and those have their MIR encoded no matter what `should_codegen` returns.

File: compiler/rustc_monomorphize/src/collector.rs [hunk 1]
```diff
@@ -203,9 +203,14 @@ use crate::errors::{
     EncounteredErrorWhileInstantiating, LargeAssignmentsLint, RecursionLimit, TypeLengthLimit,
 };
 
-#[derive(PartialEq)]
+#[derive(Debug, Copy, Clone)]
 pub enum MonoItemCollectionMode {
-    Eager,
+    Eager {
+        /// Whether to use optimized mir for collection or just analyis MIR.
```
Comment (bjorn3) [inline compiler/rustc_monomorphize/src/collector.rs:209]:
```suggestion
        /// Whether to use optimized mir for collection or just analysis MIR.
```

**Comment by Dylan-DPC** [comment]
Closing this as it was an experiment

**Pull request closed without merging**